src/Binding.lua

```diff
@@ -53,13 +66,49 @@ function bindingPrototype:map(valueTransform)
 	end
 
 	local binding = Binding.create(valueTransform(self:getValue()))
+	local internalData = binding[InternalData]
 
-	binding[InternalData].valueTransform = valueTransform
-	binding[InternalData].upstreamBinding = self
+	internalData.valueTransform = valueTransform
+
+	internalData.upstreamBindings.source = self
+	internalData.upstreamBindingCount = internalData.upstreamBindingCount + 1
 
 	return binding
 end
 
+--[[
+	Determines the final (not yet transformed) value from upstream bindings
+]]
+function bindingPrototype:__getValueFromUpstreamBindings()
+	local internalData = self[InternalData]
+	local newValue = mapBindingsToValues(internalData.upstreamBindings)
+
+	if not internalData.isJoinedBinding then
+		--[[
+			If this is not a joined binding, there will always only be one upstream
+			binding.
+
+			To ensure that joined bindings with a single upstream binding always
+			result in a table, we use the internal variable isJoinedBinding
+		]]
+		local _, value = next(newValue)
+		newValue = value
+	end
+
+	return newValue
+end
+
+--[[
+	Disconnects all connections to upstream bindings
+]]
+function bindingPrototype:__upstreamDisconnect()
+	local internalData = self[InternalData]
+
+	for _, disconnect in pairs(internalData.upstreamConnections) do
```

Comment:
We should clear out `upstreamConnections` in this method, either by emptying the array as we iterate through it or just by replacing it with a new array. I'm not sure which one is preferable in this case, or if it really matters.